BatchMmul issues follow-up

#### Issue Description

I checked out the new snapshot to try out the revised `batchMmul` op and can confirm that all the issues in #9773 are fixed. However, I have some new remarks about `batchMmul` that you might want to address to improve it:
```
INDArray a = Nd4j.ones(2, 6);
INDArray b = Nd4j.ones(6, 4);
INDArray alpha = Nd4j.scalar(1f);
INDArray beta = Nd4j.scalar(0f);
```
1. All parameters of batchMmul must have floating point types. If they are not, the validation fails. But this happens rather late (especially in Nd4j, where the exception is also a bit cryptic; see below). This doesn't necessarily need to be fixed, but I wonder why the op cannot, e.g., simply use `SDValidation.validateFloatingPoint` instead of `SDValidation.validateNumerical`.
```
INDArray[] res1 = Nd4j.base.batchMmul(alpha.castTo(DataType.INT32), beta.castTo(DataType.INT32), new INDArray[]{a}, b);
System.out.println(res1[0]);
```
2. All parameters must have the same (floating point) type. As this is only validated for `alpha` and `beta`, you may mix, e.g., `FLOAT` and `DOUBLE` data types. This leads to wrong results in both Nd4j and SameDiff.
```
INDArray[] res2 = Nd4j.base.batchMmul(alpha.castTo(DataType.DOUBLE), beta.castTo(DataType.DOUBLE), new INDArray[]{a}, b);
System.out.println(res2[0]);
/*
expected:
[[    6.0000,    6.0000,    6.0000,    6.0000], 
 [    6.0000,    6.0000,    6.0000,    6.0000]]
got:
[[ 8192.0020, 8192.0020,         0,         0], 
 [ 8192.0020, 8192.0020,         0,         0]]
*/
```
3. If I understand GEMM correctly, the formula is `alpha * A * B + beta * C`. Then the `beta` parameter of `batchMmul` seems pointless to me, as you cannot specify a matrix `C` and thus `beta` has no effect at all (in my experiments, no value I passed for `beta` changed the result). You could either remove `beta` from the API or add `C` to the API to resolve this.

#### Version Information

* Deeplearning4j version: 1.0.0-SNAPSHOT
* Platform information (OS, etc): Linux Mint 20.3
* CUDA version, if used: N/A
* NVIDIA driver version, if in use: N/A

#### Additional Information

* Exception stack trace for remark 1:
```
Op [batched_gemm] failed check for input [0], DataType: [INT32]
[main] ERROR org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner - Failed to calculate output shapes for op batched_gemm. Attempted to execute with 4 inputs, 0 outputs, 0 targs, 9 iargs, 0 bargs and 0 dargs. Inputs: [(INT,[],c), (INT,[],c), (FLOAT,[2,6],c), (FLOAT,[6,4],c)] - Please see above message (printed out from c++) for a possible cause of error.
Exception in thread "main" org.nd4j.linalg.exception.ND4JIllegalStateException: Op name batched_gemm - no output arrays were provided and calculateOutputShape failed to execute
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1513)
	at org.nd4j.linalg.factory.Nd4j.exec(Nd4j.java:6566)
	at org.nd4j.linalg.factory.ops.NDBase.batchMmul(NDBase.java:220)
	at Mmul.main(Mmul.java:32)
Caused by: java.lang.RuntimeException: Op batched_gemm with name 53b1008c-c75e-47c9-bab7-f6cfc4deaedc failed to execute. Here is the error from c++: Data types validation failed
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.calculateOutputShape(NativeOpExecutioner.java:1674)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.calculateOutputShape(NativeOpExecutioner.java:1580)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1502)
	... 4 more
```

--- a/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/memory/ArrayCacheMemoryMgr.java
+++ b/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/memory/ArrayCacheMemoryMgr.java
@@ -50,8 +50,59 @@ public class ArrayCacheMemoryMgr extends AbstractMemoryMgr {
 
     private static Map<INDArray,INDArray> released = new IdentityHashMap<>();
 
+    @Getter
+    @Setter
     private static  AtomicDouble maxMemFrac;
+   @Getter
+   @Setter
     private static AtomicLong smallArrayThreshold;
+
+    public static Set<Long> getLruCache() {
+        return lruCache;
+    }
+
+    public static Map<Long, INDArray> getLruCacheValues() {
+        return lruCacheValues;
+    }
+
+    public static AtomicDouble getMaxMemFrac() {
+        return maxMemFrac;
+    }
+
+    public static void setMaxMemFrac(AtomicDouble maxMemFrac) {
+        ArrayCacheMemoryMgr.maxMemFrac = maxMemFrac;
+    }
+
+    public static void setMaxMemFrac(double maxMemFrac) {
+        ArrayCacheMemoryMgr.maxMemFrac.set(maxMemFrac);
+    }
+
+    public static AtomicDouble getLargerArrayMaxMultiple() {
+        return largerArrayMaxMultiple;
+    }
+
+    public static void setLargerArrayMaxMultiple(AtomicDouble largerArrayMaxMultiple) {
+        ArrayCacheMemoryMgr.largerArrayMaxMultiple = largerArrayMaxMultiple;
+    }
+    public static void setLargerArrayMaxMultiple(double largerArrayMaxMultiple) {
+        ArrayCacheMemoryMgr.largerArrayMaxMultiple.set(largerArrayMaxMultiple);
+    }
+    public static AtomicLong getMaxCacheBytes() {
+        return maxCacheBytes;
+    }
+
+    public static void setMaxCacheBytes(AtomicLong maxCacheBytes) {
+        ArrayCacheMemoryMgr.maxCacheBytes = maxCacheBytes;
+    }
+
+    public static AtomicLong getCurrentCacheSize() {
+        return currentCacheSize;
+    }
+
+    public static void setCurrentCacheSize(AtomicLong currentCacheSize) {
+        ArrayCacheMemoryMgr.currentCacheSize = currentCacheSize;
+    }
+
     private static AtomicDouble largerArrayMaxMultiple;
 
     private static AtomicLong maxCacheBytes;

--- a/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Gather.java
+++ b/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Gather.java
@@ -165,6 +165,7 @@ public class Gather extends DynamicCustomOp {
     public List<SDVariable> doDiff(List<SDVariable> i_v) {
         //2 args: input and indices. Plus integer dimension arg
         //Gather backprop is just scatter add
+        //credit for fixes to @paratsu
         SDVariable indicesGrad = sameDiff.zerosLike(arg(1));
         var gradAtOut = i_v.get(0);
         try (var ignore = sameDiff.withNameScope(gradAtOut.name())) {

--- a/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Gather.java
+++ b/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Gather.java
@@ -38,12 +38,14 @@ import org.tensorflow.framework.NodeDef;
 
 import java.util.*;
 
+import static org.nd4j.linalg.api.buffer.DataType.INT32;
+
 /**
  * Gather op
  */
 public class Gather extends DynamicCustomOp {
 
-    protected int[] indices;
+    protected int[] indices = null;
     protected int jaxis = 0;
 
     public Gather() {
@@ -163,68 +165,38 @@ public class Gather extends DynamicCustomOp {
     public List<SDVariable> doDiff(List<SDVariable> i_v) {
         //2 args: input and indices. Plus integer dimension arg
         //Gather backprop is just scatter add
-        SDVariable indicesSize = sameDiff.expandDims(args()[1].length(),0);
-        SDVariable paramsShape = sameDiff.shape(args()[0]);
-        paramsShape = paramsShape.reshape(paramsShape.length());
         SDVariable indicesGrad = sameDiff.zerosLike(arg(1));
+        var gradAtOut = i_v.get(0);
+        try (var ignore = sameDiff.withNameScope(gradAtOut.name())) {
+            //Gather backprop is just scatter add
+            SDVariable inputArray = arg(0);
+            SDVariable indices = args().length > 1 ? arg(1) : sameDiff.constant(Nd4j.createFromArray(this.indices));
+            SDVariable inputGrad = sameDiff.zerosLike(inputArray);
+            SDVariable inputArrayRank = inputArray.rank();
+            SDVariable gatherAxis = (jaxis < 0 ? inputArrayRank.minus(1) : sameDiff.constant(jaxis)).reshape(1);
+            SDVariable gradAtOutAdditionalDimensions = sameDiff.range(inputArrayRank, gradAtOut.rank(), sameDiff.constant(1), INT32);
 
-        if(jaxis == 0) {
-            SDVariable paramsTailShape = paramsShape.getView(SDIndex.interval(sameDiff.constant(1)
-                    , sameDiff.constant(1),paramsShape.length()));
-            SDVariable valueShape = sameDiff.concat(0,indicesSize,paramsTailShape);
-            SDVariable values = sameDiff.reshape(i_v.get(0),valueShape);
-            SDVariable indices = sameDiff.flatten(args()[1]);
-            SDVariable retGrad = sameDiff.zerosLike(arg());
-            SDVariable put = retGrad.put(indices,values,indices).reshape(arg().shape());
-            /**
-             * TODO: figure out a better way to do a mass assign.
-             * We can't match the speed of a sparse gradient so we need to figure out the best way to
-             * achieve this with a dense representation.
-             *
-             * This would ideally be similar to nd4j's put(indices)
-             */
-            return Arrays.asList(put, indicesGrad);
-        } else {
-            SDVariable batchDims = sameDiff.constant(0);
-            SDVariable outerShape = paramsShape.getView(SDIndex.interval(0,jaxis));
-            SDVariable innerShape = paramsShape.getView(
-                    SDIndex.interval(sameDiff.constant(jaxis),paramsShape.length()),SDIndex.interval(sameDiff.constant(1),sameDiff.constant(-1)));
-            SDVariable valueShape = sameDiff.concat(0,outerShape,
-                    sameDiff.constant(-1).castTo(outerShape.dataType()),
-                    innerShape.castTo(outerShape.dataType()));
+            //Use scatter add plus permute
+            SDVariable inputArrayDimensions = sameDiff.range(null, sameDiff.constant(0), inputArrayRank, sameDiff.constant(1), INT32);
+            SDVariable inputArrayDimensionsRectified =
+                    sameDiff.math().listDiff(inputArrayDimensions, gatherAxis)[0];
 
+            // Indices
+            SDVariable inputPermuteDims = sameDiff.concat(0, gatherAxis, inputArrayDimensionsRectified);
+            SDVariable outGradPermuteDims =
+                    sameDiff.concat(0, inputPermuteDims, gradAtOutAdditionalDimensions);
+            SDVariable inputInvertDims = sameDiff.invertPermutation(inputPermuteDims);
 
-            /**
-             * Blow grad up to match values shape, values shape is  not wrong
-             */
-            SDVariable valuesDims = valueShape.length();
-            SDVariable axisDims = outerShape.length();
+            //Permute gradients so original axis is at position 0... then scatter add, and reverse
+            SDVariable permutedOutGrad = gradAtOut.permute(outGradPermuteDims);
+            SDVariable inputGradPermuted =inputGrad.permute(inputPermuteDims);
+            SDVariable inputGradPermutedScatterSum = sameDiff.scatterAdd(inputGradPermuted, indices, permutedOutGrad);
 
-            SDVariable outerBatchIndices = sameDiff.range(0,0,0,DataType.INT64);
-            SDVariable batchAxisIndices = sameDiff.range(batchDims,axisDims, sameDiff.constant(1),DataType.INT64);
-            SDVariable innerAxisIndices = sameDiff.range(axisDims.add(1.0),valuesDims,sameDiff.constant(1),DataType.INT64);
-
-            SDVariable indices = sameDiff.reshape(args()[1],indicesSize);
-
-            SDVariable put = sameDiff.unsortedSegmentSum(i_v.get(0), sameDiff.range(sameDiff.constant(0),sameDiff.sizeAt(i_v.get(0),0),sameDiff.constant(1),DataType.INT64), sameDiff.sizeAt(i_v.get(0),0));
-            SDVariable values = sameDiff.reshape(put,valueShape);
-
-
-
-            SDVariable transposeDims = sameDiff.concat("transposeConcat",0,outerBatchIndices,axisDims,batchAxisIndices,innerAxisIndices);
-            SDVariable valuesTranspose = sameDiff.permute(values,transposeDims);
-
-            /**
-             * Batch gather grad
-             */
-
-            SDVariable paramsGrad = sameDiff.unsortedSegmentSum(valuesTranspose,indices,paramsShape.get(SDIndex.point(jaxis)));
-            SDVariable invertTransposeDims = sameDiff.concat(0,outerBatchIndices.castTo(DataType.INT64),batchAxisIndices.add(1).castTo(DataType.INT64),batchDims.castTo(DataType.INT64),innerAxisIndices.castTo(DataType.INT64));
-            paramsGrad = sameDiff.permute(paramsGrad,invertTransposeDims);
-
-
-            return Arrays.asList(paramsGrad, indicesGrad);
+            //Now, invert the permutation so axis is back where it was
+            SDVariable finalInputGrad = inputGradPermutedScatterSum.permute(inputInvertDims);
+            return Arrays.asList(finalInputGrad,indicesGrad);
         }
+
 
     }
 

--- a/platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/autodiff/samediff/MemoryMgrTest.java
+++ b/platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/autodiff/samediff/MemoryMgrTest.java
@@ -33,6 +33,7 @@ import org.nd4j.linalg.factory.Nd4j;
 import org.nd4j.linalg.factory.Nd4jBackend;
 
 import java.lang.reflect.Field;
+import java.util.concurrent.atomic.AtomicLong;
 
 import static org.junit.jupiter.api.Assertions.*;
 
@@ -52,7 +53,7 @@ public class MemoryMgrTest extends BaseNd4jTestWithBackends {
     public void testArrayReuseTooLarge(Nd4jBackend backend) throws Exception {
 
         ArrayCacheMemoryMgr mmgr = new ArrayCacheMemoryMgr();
-        mmgr.setMaxCacheBytes(1000);
+        mmgr.setMaxCacheBytes(new AtomicLong(1000));
         assertEquals(1000, mmgr.getMaxCacheBytes());
 
         INDArray[] arrays = new INDArray[100];

